### ISSUE TITLE
Inconsistent results when scanning with rootfs

### ISSUE THREAD

**Issue by vishiy** [aquasecurity/trivy#2037]
## Description

We have an image that we unpack and scan with rootfs  as specified here - https://aquasecurity.github.io/trivy/v0.20.0/advanced/container/unpacked-filesystem/

Trivy version v0.26.0

docker export $(some image) | tar -C ./trivytemp -xvf -

1) I am scanning rootfs pointing to the root (./trivytemp), looking for a sample CVE that exists on a ruby gem

`trivy rootfs ./trivytemp/usr/bin | grep bundler`     this returns nothing

2) now i scan rootfs pointing to specific path under rootfs (in this case ./trivitemp/usr/lib)  
`trivy rootfs ./trivytemp/usr/lib | grep bundler`     this returns CVEs as expected

| bundler                        | CVE-2020-36327   | HIGH     | 2.1.4             | >= 2.2.18, 2.2.10            | rubygem-bundler: Dependencies         |
| (bundler-2.1.4.gemspec)        |                  |          |                   |                              | of gems with explicit source          |
|                                | CVE-2021-43809   | MEDIUM   |                   | >= 2.2.33                    | rubygem-bundler: unexpected           |

Why this discrepancy ? So is it expected to scan every possible path under root when scanning with rootfs ? Seems not right.

Expect `rootfs` scans all paths recursively, but it seems to be skipping some paths.

**Comment by knqyf263** [comment]
First of all, you are not scanning `./trivytemp`.

>trivy rootfs ./trivytemp/usr/bin | grep bundler this returns nothing

Even if you tried `./trivytemp` and it returned nothing, it may be related to system-installed packages. How was bundler installed? If it comes from a package manager such as `yum` and `apt`, Trivy uses security advisories from vendors such as Red Hat and Debian, not from the upstream because the vendor fixed version is different from the upstream.

Trivy can detect such system installed packages by scanning root filesystem, but cannot detect it if you pass the specific path such as `usr/lib`. I'd recommend checking how bundler is installed in your image and seeing security advisories from vendors if it is a system-installed package. If you confirm it was installed via `gem` and still see the discrepancy, please let us know.

**Comment by vishiy** [comment]
@knqyf263 - i am seeing the opposite. when i scan "/' with rootfs option, i dont get certain CVEs, which i get them when i scan with specific pathy in the file system with rootfs option ex;- /usr/lib. Is rootfs is doing the same thing when scanning the root v a sub path ? it seems not.

**Comment by knqyf263** [comment]
> i am seeing the opposite. when i scan "/' with rootfs option, i dont get certain CVEs

This is exactly what I said. Let's say an upstream of bundler has a patch in v1.1.2 and a vendor backports it to v1.0.1-1. There are two security advisories for the same vulnerability. We need to know which advisory should be applied, otherwise, it leads to false detection. If your installed bundler is v1.0.2, nobody knows whether it is vulnerable or not until we know where it comes from. If it is from upstream, it is vulnerable as v1.0.2 < v1.1.2 obviously. On the other hand, if it is from the vendor, it is not vulnerable as v1.0.2 > v1.0.1-1.

>is rootfs doing the same thing when scanning the root vs. scanning a specific sub path ? it seems not.

They are different since scanning root dir can detect system-installed packages, then try to use vendor advisories. The system info is usually under `/var/lib/dpkg/info/`. If you pass `./trivytemp`, the info will be found as it exists in `./trivytemp/var/lib/dpkg/info/`. Therefore, Trivy can know which packages are installed by `apt`. But if you pass `./trivytemp/usr/lib/`, no info is found in `./trivytemp/usr/lib/var/lib/dpkg/info`. Then, Trivy needs to assume `bundler` is installed from upstream.

In general, the difference totally makes sense. But in the case of your image, there is another thing. The detected bundler is installed via `libruby2.7`.

```
root@357b2c9607e8:/opt# dpkg -L libruby2.7 | grep bundler-2.1.4.gemspec
/usr/lib/ruby/gems/2.7.0/specifications/default/bundler-2.1.4.gemspec
```

It is obviously a package installed by `apt`, so we cannot use the upstream advisory like [GHSA](https://github.com/advisories/GHSA-fj7f-vq84-fh43). We have to use a security advisory from Ubuntu as your image is Ubuntu 18.04. Let's see [the Ubuntu advisory](https://ubuntu.com/security/CVE-2021-43809). A package mentioned is only `bundler`, but your image doesn't have the `bundler` package.

```
root@357b2c9607e8:/opt# dpkg -L bundler
dpkg-query: package 'bundler' is not installed
```

Also,  `libruby2.7` is not listed even though `bundler-2.1.4` is installed via `libruby2.7`. There are two possibilities. 

1. `libruby2.7` is not affected by CVE-2021-43809
2. Ubuntu just forgot to add `libruby2.7` in the list

From the perspective of a security scanner, we should rely on the Ubuntu advisory if the package is installed by `apt`. If you think their advisory is wrong, we can report it to Ubuntu.

In short, the different behavior is intended because of system info detection.
